micro_sam/training/joint_sam_trainer.py

```diff
@@ -0,0 +1,161 @@
+
+    def _instance_train_iteration(self, x, y):
+        outputs = self.unetr(x.to(self.device))
+        loss = self.instance_loss(outputs, y.to(self.device))
+        return loss
+
+    def _train_epoch_impl(self, progress, forward_context, backprop):
+        self.model.train()
+
+        input_check_done = False
+
+        n_iter = 0
+        t_per_iter = time.time()
+        for x, y in self.train_loader:
+            labels_instances = y[:, 0, ...].unsqueeze(1)
+            labels_for_unetr = y[:, 1:, ...]
+
+            input_check_done = self._check_input_normalization(x, input_check_done)
+
+            self.optimizer.zero_grad()
+
+            with forward_context():
+                # 1. train for the interactive segmentation
+                (loss, mask_loss, iou_regression_loss, model_iou,
+                 sampled_binary_y) = self._interactive_train_iteration(x, labels_instances)
+
+            backprop(loss)
+
+            self.optimizer.zero_grad()
+
+            with forward_context():
+                # 2. train for the automatic instance segmentation
+                unetr_loss = self._instance_train_iteration(x, labels_for_unetr)
+
+            backprop(unetr_loss)
+
+            if self.logger is not None:
+                lr = [pm["lr"] for pm in self.optimizer.param_groups][0]
+                samples = sampled_binary_y if self._iteration % self.log_image_interval == 0 else None
+                self.logger.log_train(
+                    self._iteration, loss, lr, x, labels_instances, samples,
+                    mask_loss, iou_regression_loss, model_iou, unetr_loss
+                )
+
+            self._iteration += 1
+            n_iter += 1
+            if self._iteration >= self.max_iteration:
+                break
+            progress.update(1)
+
+        t_per_iter = (time.time() - t_per_iter) / n_iter
+        return t_per_iter
+
+    def _validate_impl(self, forward_context):
+        self.model.eval()
+
+        input_check_done = False
+
+        val_iteration = 0
+        metric_val, loss_val, model_iou_val = 0.0, 0.0, 0.0
+
+        with torch.no_grad():
+            for x, y in self.val_loader:
+                labels_instances = y[:, 0, ...].unsqueeze(1)
+                labels_for_unetr = y[:, 1:, ...]
+
+                input_check_done = self._check_input_normalization(x, input_check_done)
+
+                with forward_context():
+                    # 1. validate for the interactive segmentation
+                    (loss, mask_loss, iou_regression_loss, model_iou,
+                     sampled_binary_y, metric) = self._interactive_val_iteration(x, labels_instances, val_iteration)
+
+                with forward_context():
+                    # 2. validate for the automatic instance segmentation
+                    unetr_loss = self._instance_train_iteration(x, labels_for_unetr)
+
+                loss_val += loss.item()
+                metric_val += metric.item()
```

Comment:
Potentially we could also add the unetr loss as metric here. (Let's discuss tomorrow.)

Reply:
Yes, that's something I wanted to discuss. Let's check in tomorrow. Thanks!

Reply:
This is taken care of now. Thanks!